src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -227,18 +500,200 @@ citus_internal_unregister_tenant_schema_globally(PG_FUNCTION_ARGS)
+	ExecuteForeignKeyCreateCommandList(originalForeignKeyRecreationCommands,
+									   skip_validation);
+
+	/* Register the schema locally and sync it to workers */
+	InsertTenantSchemaLocally(schemaId, colocationId);
+	char *registerSchemaCommand = TenantSchemaInsertCommand(schemaId, colocationId);
+	if (EnableMetadataSync)
+	{
+		SendCommandToWorkersWithMetadata(registerSchemaCommand);
+	}
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * citus_schema_undistribute gets a tenant schema name, then converts it to a regular
+ * schema by undistributing all tables under it.
+ */
+Datum
+citus_schema_undistribute(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid schemaId = PG_GETARG_OID(0);
+	EnsureSchemaExist(schemaId);
+	EnsureSchemaOwner(schemaId);
+
+	/* Prevent concurrent table creation under the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
+
+	/*
+	 * We should ensure the existence of the schema after taking the lock since
+	 * the schema could have been dropped before we acquired the lock.
+	 */
+	EnsureSchemaExist(schemaId);
+	EnsureSchemaOwner(schemaId);
+
+	/* The schema should be a tenant schema */
+	char *schemaName = get_namespace_name(schemaId);
+	if (!IsTenantSchema(schemaId))
+	{
+		ereport(ERROR, (errmsg("schema %s is not distributed", schemaName)));
+	}
+
+	ereport(NOTICE, (errmsg("undistributing schema %s", schemaName)));
+
+	/* Take lock on the relations and filter out partition tables */
+	List *tableIdListInSchema = GetRegularAndForeignCitusTablesInSchema(schemaId);
+	List *tableIdListToConvert = NIL;
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, tableIdListInSchema)
+	{
+		/* prevent concurrent drop of the relation */
+		LockRelationOid(relationId, AccessShareLock);
+		EnsureRelationExists(relationId);
+
+		/*
+		 * Skip partitions as they would be undistributed by the parent table.
+		 *
+		 * We should filter out partitions here before undistributing the schema.
+		 * Otherwise, converted partitioned table would change oid of partitions and its
+		 * partition tables would fail with oid not exist.
+		 */
+		if (PartitionTable(relationId))
+		{
+			continue;
+		}
+
+		tableIdListToConvert = lappend_oid(tableIdListToConvert, relationId);
+
+		/* Only single shard tables are expected during the undistribution of the schema */
+		Assert(IsCitusTableType(relationId, SINGLE_SHARD_DISTRIBUTED));
+	}
+
+	/*
+	 * First, we need to delete schema metadata and sync it to workers. Otherwise,
+	 * we would get error from `ErrorIfTenantTable` while undistributing the tables.
+	 */
+	char *unregisterSchemaCommand = TenantSchemaDeleteCommand(schemaName);
+	DeleteTenantSchemaLocally(schemaId);
+	if (EnableMetadataSync)
+	{
+		SendCommandToWorkersWithMetadata(unregisterSchemaCommand);
+	}
+
+	/*
+	 * No need to drop colocation entries explicitly here, since colocation entry would be
+	 * automatically dropped when we undistribute the last table in the schema.
+	 */
```

Comment:
I know I asked dropping DeleteColocationGroup bits from here but apperantly this causes leaking the colocation group if the tenant schema is empty.

We should probably save the colocation id associated to tenant schema, before TenantSchemaDeleteCommand, and need to use it here to perform DeleteColocationGroup related stuff:

```sql
\c - onurctirtir

CREATE SCHEMA s6;

CREATE SCHEMA s7;
CREATE TABLE s7.t1(a int PRIMARY KEY, b int, c text);

ALTER SCHEMA s6 OWNER TO test_non_super_user;
ALTER SCHEMA s7 OWNER TO test_non_super_user;

SELECT run_command_on_workers($$ALTER SCHEMA s6 OWNER TO test_non_super_user;$$);
SELECT run_command_on_workers($$ALTER SCHEMA s7 OWNER TO test_non_super_user;$$);

ALTER TABLE s7.t1 OWNER TO test_non_super_user;

\c - test_non_super_user

select citus_schema_distribute('s6');
select citus_schema_distribute('s7');

-- check pg_dist_tenant_schema & pg_dist_colocation
-- they look sane

select citus_schema_undistribute('s6');
select citus_schema_undistribute('s7');
```

```sql
-- check pg_dist_tenant_schema & pg_dist_colocation
-- colocation id for s6 is not deleted from pg_dist_colocation
```

Reply:
We might want to move the last 4 lines of citus_internal_unregister_tenant_schema_globally() into an helper function and use it here too:

```c
UnregisterTenantSchemaGlobally()
	uint32 tenantSchemaColocationId = SchemaIdGetTenantColocationId(schemaId);

	DeleteTenantSchemaLocally(schemaId);
	SendCommandToWorkersWithMetadata(TenantSchemaDeleteCommand(schemaNameStr));

	DeleteColocationGroup(tenantSchemaColocationId);
```